demos/enclave/ETL/compile_models.py

```diff
@@ -0,0 +1,335 @@
+            "email":
+                oneroster_users["email"][i].lower()
+        }
+        users.append(User.parse_obj(data))
+    return pd.DataFrame([s.__dict__ for s in users])
+
+
+def demographics_model(oneroster_demographics, sourceid_2_email, email_2_uuid, ):
+    demographics = []
+    for i in range(len(oneroster_demographics["sourceId"])):
+        uuid = email_2_uuid[sourceid_2_email[oneroster_demographics["sourceId"][i]]]
+        data = {
+            "user_uuid":
+                uuid,
+            "birth_date":
+                oneroster_demographics["birthDate"][i],
+            "sex":
+                oneroster_demographics["sex"][i],
+            "american_indian_or_alaska_native":
+                str(oneroster_demographics["americanIndianOrAlaskaNative"][i]).lower(),
+            "asian":
+                str(oneroster_demographics["asian"][i]).lower(),
+            "black_or_african_american":
+                str(oneroster_demographics["blackOrAfricanAmerican"][i]).lower(),
+            "native_hawaiian_or_other_pacific_islander":
+                str(oneroster_demographics["naitiveHawaiianOrPacificIslander"][i]).lower(),
+            "white":
+                str(oneroster_demographics["white"][i]).lower(),
+            "demographic_race_two_or_more_races":
+                str(oneroster_demographics["demographicRaceTwoOrMoreRaces"][i]).lower(),
+            "hispanic_or_latino_ethnicity":
+                str(oneroster_demographics["hispanicOrLatinoEthnicity"][i]).lower()
+        }
+        demographics.append(Demographic.parse_obj(data))
+    return pd.DataFrame([s.__dict__ for s in demographics])
+
+
+def make_name2assessmentID(cli_grades):
+    mapping = {}
+    count = 0
+    for classroom in cli_grades.keys():
+        for user_item in cli_grades[classroom]["usergrades"]:
+            for grade_item in user_item["gradeitems"]:
+                if grade_item["itemname"] is not None:
+                    name = grade_item["itemname"].strip('"')
+                    if name not in mapping.keys():
+                        mapping[name] = count
+                        count += 1
+    return mapping
+
+def make_email2uuid(cli_users):
+    mapping = {}
+    for classroom in cli_users.keys():
+        for user in cli_users[classroom]:
+            mapping[user["email"]] = uuid.uuid4()
+    return mapping 
+
+def make_userid2email(cli_users):
+    mapping = {}
+    for classroom in cli_users.keys():
+        for user in cli_users[classroom]:
+            mapping[user["id"]] = user["email"]
+    return mapping
+
+def make_sourceid2email(oneroster_users):
+    mapping = {}
+    for i in range(0, len(oneroster_users['sourceId'])):
+        if oneroster_users["sourceId"][i] != "nan":            
+            mapping[oneroster_users["sourceId"][i]] = oneroster_users["email"][i]
+    return mapping
+
+
+def create_models(
+    output_path, cli_grades, cli_users, oneroster_demographics, oneroster_users, 
+):
```

Comment:
Putting this comment here though I think it's an overall structural question: I have a hard time following the various shuffling between dataframes to dicts to dataframes ... :sweat_smile: . For example, you're parsing the one roster data as dfs, converting to dicts, then at some point back to dfs and serialized out. I also think that some of the data mangling is probably easier to do in dataframes than Python dicts (maybe?). Also, the validation via pydantic models is sort of hidden in the mix.

I'm wondering if it might be possible to do something slightly different structurally: Since we're going to tabular forms anyways, instead of converting CSV data from dataframes to dicts, could we instead convert the relevant parts of the JSON datasets into dataframes? If so, we can use all of the manipulation capabilities from the `pandas` API (which is typically easier to read) to create the dfs that align with the schema. Then, before serializing we can just call specific validators that run through the rows of the data and make sure there are no errors if we instantiate pydantic models out of them.

Conversely, if this isn't possible or we prefer the dict based manipulation, I'm not sure why the various model functions return dataframes versus just a list of dicts that can be serialized via `DictWriter` :man_shrugging: . It seems odd to just create dataframes to write out CSVs when that could be done without the transformation.

Reply:
Darn, I was very close to asking you about this whole approach last week before I wrote it 😓. But oh well good learning moment.

So for the first approach you mention, when converting JSONs to dataframes - do you anticipate each JSON composing 2 or 3 dataframes since there are many objects within objects? 

Secondly, when you talk about the manipulation capabilities of pandas to create dfs that align with the schema, do you just mean accessing the appropriate data from various dataframes and compiling it together? If so I think I know what you are envisioning here. 

- But I could also see it being easy to keep the JSON data as a dict, the csv data as a dataframe, and then access each respectively. Or does having everything in a similar structure just make the model compilation that much simpler?

Reply:
To your last point: I think trying to keep two different types is going to be painful + hard to reason about. I would definitely pick one. Basically think of it has a pipeline like:

```
raw data format (CSV / JSON) -> internal data format -> output data format (CSV)
```

I would argue we should have one internal data format, whether that's dicts or dfs. Since we have two options, it's about which we think is more advantageous. I think dfs provide more expressive APIs for manipulation / querying (it's what they're designed for), and since we know we're producing CSVs which align well with dfs (versus if we were producing JSON / nested data structures) it seems like the more natural of the two :man_shrugging: .

In terms of manipulation capabilities, I meant things like doing joins (e.g. to get something that maps UUIDs <-> emails <-> ... using dfs versus creating the various mapper dicts), or even [renaming columns](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html) where there's a more expressive API versus doing manual manipulation of dicts in Python. Another example could be where you lower the values in user email addresses, if that was done as part of a series of steps you could use the [apply API](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html) where that would be clear we're doing some data cleaning versus sort of hidden in the transformation.

I haven't thought too hard about how many dataframes would need to be built up, but I would imagine there would be some 1:N mapping between any JSON inputs and intermediary / final dataframes for processing.